### PR TITLE
 Add ability to select path segments using Python re regexprs;  follow on to PR#146 

### PR DESCRIPTION
**Pull request by AlainLich**
Hi,  this is a follow-on of my previous PR (#146) on same subject; apparently it was simpler for me to use a different branch.

This has run Action https://github.com/AlainLich/dpath-python/actions/runs/3684590771 in my branch `AL-master4merge`. I have added tests exercising this with most functions, and documented in `README.rst`

All details are in the list of commits, including documentation, tests,...

I hope this suits your needs, let me know if there are still issues for merging in the main baseline.

### CODE REVIEW

**Comment by AlainLich** [comment]
Hi @moomoohk 

- after looking at yesterday's tests, failure was apparently caused by lack of support for typing.Protocol, which implements https://peps.python.org/pep-0544/ , because pypy's version was older than the one I have been using. 

- I am preparing an improved version that supports **both** PEP-544 compliant and ignorant compilers, with a little bit more work for the user in the latter case. 

- Apparently most of the work TBD is in the testing, so don't expect a revised version before next week-end.

**Comment by moomoohk** [comment]
Hi Alain,
I've cleared enough from the backlog and now I'd like to focus on your PR.
After reviewing the changes you've made it's clear a lot of work has been put in and that's much appreciated.
That said, I feel that a lot of the changes included aren't within the scope of the feature addition you're proposing.

If the core functionality (adding regexes to paths) works well enough then I'd like to restructure this PR. I hope that's ok with you!

Perhaps the other changes can be considered separately.

Thanks.

**Comment by AlainLich** [comment]
Hi,
 a) I will be pushing a version that cleans up the functionality you
probably are not interested in (using a class or a function to define
  matching).
 b) doing this, I have cleared the issues with running on (slightly) older
versions of the Python or Pypy (which was more difficult/surprising).
 c) having a lot of tests helped, especially because the type
identification for instances is subtle.

  So you should probably start with this forthcoming version, to clean up
and see what  you want to keep.

  For me, it is OK that you keep only the functionality you are interested
in.
  ( Personally, I got interested in the example with Levenshtein
approximate matching, when it seemed clear that this would be such a small
change)

 Looking forwards to reading you

**Comment by AlainLich** [comment]
Hi @moomoohk,
After thinking about it, I think it will be simpler for you if I committed:
a) a simplified version with only re regexp in the branch linked to the PR.
This will keep only that functionality plus workarounds for working Pypy if
required.
b) another full featured version in a new branch, which includes what I
have now and can be readied for a PR after some testing in Github's
containers

This plan should be simpler for you, since I know what I put in!

Have a good day !😀

**Comment by AlainLich** [comment]
Hi @moomoohk 

a) this is fine, do you expect me **to do** anything?
b) while you are on this, could you make your workflow "deploy" dependent on your user name; the way things are set up now, each time I pull master for synchronising (and if the version has changed), it launches (under my name) and
fails as it should... and I get some spurious mail. ( https://github.com/AlainLich/dpath-python/actions/runs/3900678773)

Something like the following should work (not tried):
```
deploy:
   ... 
    # Check that this only occurs on maintainer repo
    if: ${{ github.actor == "dpath-maintainers" }} 

```

Have a good day :-)

File: dpath/options.py [hunk 1]
```diff
@@ -1 +1,38 @@
...
+
+# --------------------------------------------------------------------
+# Language processor variations
+# --------------------------------------------------------------------
+
+# Using Unions of types in isinstance is also an issue
+# https://peps.python.org/pep-0604/#isinstance-and-issubclass
+# https://bugs.python.org/issue44529
+
+try:
```
Comment (moomoohk) [inline dpath/options.py:29]: What is this check needed for?

File: dpath/segments.py [hunk 2]
```diff
@@ -6,6 +6,12 @@
 from dpath.exceptions import InvalidGlob, InvalidKeyName, PathNotFound
 from dpath.types import PathSegment, Creator, Hints, Glob, Path, SymmetricInt
 
+import re
+try:
```
Comment (moomoohk) [inline dpath/segments.py:10]: Dpath's minimum supported version is Python 3.7 making this check redundant

**Comment by moomoohk** [comment]
Hi Alain

> Hi @moomoohk
> 
> a) this is fine, do you expect me **to do** anything? b) while you are on this, could you make your workflow "deploy" dependent on your user name; the way things are set up now, each time I pull master for synchronising (and if the version has changed), it launches (under my name) and fails as it should... and I get some spurious mail. ( https://github.com/AlainLich/dpath-python/actions/runs/3900678773)
> 
> Something like the following should work (not tried):
> 
> ```
> deploy:
>    ... 
>     # Check that this only occurs on maintainer repo
>     if: ${{ github.actor == "dpath-maintainers" }} 
> ```
> 
> Have a good day :-)

Hi Alain,
I might have some time to go through your changes over the next few days.
I would much appreciate if you could open a separate PR/restructure this one to include the necessary changes for your proposed features.
I'll be glad to review miscellaneous infrastructure changes in a separate PR.
As it stands, the miscellaneous changes are blocking your feature.

Thanks and regards

**Comment by AlainLich** [comment]
Hi @moomoohk,
I will try to look into this before the end of next week-end. I will send you a few questions about what to keep in the mean time. Will keep the bare minimum for the feature to work, and for tests, and prepare a PR based on current master commit.

For now:
a) do you want to always enable the feature ? to allow enabling/disablig by `DPATH_ACCEPT_RE_REGEXP`  ?
b) I will review tests I have added for their usefulness

Regards

**Comment by AlainLich** [comment]
Hi @moomoohk 
following your remarks, I submitted PR #186 as replacement for this one, with only required changes.  This PR #177 can be closed.

You were right to spot not needed stuff, some that can be assumed for Python >= 3.7, others that were left from my exercise with duck-typed generalization. 

Regards
Alain

**Comment by moomoohk** [comment]
May we close this PR for the time being?

**Pull request closed without merging**